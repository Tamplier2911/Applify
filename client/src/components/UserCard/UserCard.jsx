// import "./UserCard.scss";
import React from "react";

// redux
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { selectCurrentLanguage } from "../../redux/lang/lang.selectors";
import { deleteOneUserStart } from "../../redux/users/users.actions";

// components
import Button from "../Button/Button";

// data formaters
import getImageRelativePath from "../../utils/PathTransformations/getImageRelativePath";
import transformDateToLocaleString from "../../utils/DataTransformations/transformDateToLocaleString";

// JS Rendering CSS
import {
  UserCardContainer,
  UserCardHeader,
  UserCardHeaderName,
  UserCardHeaderEmail,
  UserCardHeaderRole,
  UserCardBody,
  UserCardBodyWrap,
  UserCardBodyImg,
  UserCardBodyAbout,
  UserCardBodyAboutLQT,
  UserCardBodyAboutCTN,
  UserCardBodyAboutRQT,
  UserCardBodyPassword,
  UserCardBodyPasswordOptions,
  UserCardBodyPasswordChanged,
  UserCardControlls,
  UserCardControllsLink,
} from "./UserCardStyles";

// component constants
import userCardData from "./UserCardConstants";

const UserCard = ({ data, lang, deleteOneUserStart }) => {
  const { _id, photo, about, role, name, email, passwordChangedAt } = data
    ? data
    : {};

  const long = about.slice(0, 80)[79];
  const shortenAbout = about.slice(0, 80);
  const image = getImageRelativePath(photo ? photo : "");
  const date = transformDateToLocaleString(
    passwordChangedAt ? passwordChangedAt : ""
  );

  const {
    userCardDelete,
    userCardUpdate,
    userCardAbout,
    userCardPasswordChangedAt,
    userCardPasswordNotYet,
  } = userCardData[lang];

  return (
    <UserCardContainer>
      <UserCardHeader>
        <UserCardHeaderName>{name}</UserCardHeaderName>
        <UserCardHeaderEmail>{email}</UserCardHeaderEmail>
        <UserCardHeaderRole>{role}</UserCardHeaderRole>
      </UserCardHeader>
      <UserCardBody>
        <UserCardBodyWrap>
          <UserCardBodyImg src={image} alt="happy user" />
        </UserCardBodyWrap>
        <UserCardBodyAbout>
          <UserCardBodyAboutLQT>&ldquo;</UserCardBodyAboutLQT>
          <UserCardBodyAboutCTN>
            {about && long
              ? shortenAbout + "..."
              : about
              ? about
              : userCardAbout}
          </UserCardBodyAboutCTN>
          <UserCardBodyAboutRQT>&bdquo;</UserCardBodyAboutRQT>
        </UserCardBodyAbout>
        <UserCardBodyPassword>
          <UserCardBodyPasswordOptions>
            {userCardPasswordChangedAt}
          </UserCardBodyPasswordOptions>
          <UserCardBodyPasswordChanged>
            {passwordChangedAt ? date : userCardPasswordNotYet}
          </UserCardBodyPasswordChanged>
        </UserCardBodyPassword>
      </UserCardBody>
      <UserCardControlls>
        <Button
          type="button"
          value={userCardDelete}
          click={() => deleteOneUserStart(_id)}
        />
        <UserCardControllsLink to={`/profile/users/${_id}`}>
          {userCardUpdate}
        </UserCardControllsLink>
      </UserCardControlls>
    </UserCardContainer>
  );
};

const mapStateToProps = createStructuredSelector({
  lang: selectCurrentLanguage,
});

export default connect(mapStateToProps, { deleteOneUserStart })(UserCard);
