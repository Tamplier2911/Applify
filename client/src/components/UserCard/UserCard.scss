/*

.user-card {
  display: grid;
  grid-template-rows: 5fr 5fr 1fr;
  grid-row-gap: 1rem;
  font-family: var(--font-title);
  padding: 2rem;
  box-shadow: 0rem 0rem 0.8rem var(--cl-font);

  transition: box-shadow 0.3s;

  &:hover {
    box-shadow: 0rem 0rem 0.8rem var(--cl-primary);
  }

  &__header {
    display: grid;
    grid-auto-rows: min-content;

    &--name {
      text-align: center;
      font-size: 1.8rem;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }
    &--email {
      text-align: center;
      font-size: 1.8rem;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }
    &--role {
      text-align: center;
      font-size: 1.8;
    }
  }
  &__body {
    position: relative;
    display: grid;
    grid-template-rows: min-content 1fr;

    &--imgWrap {
      position: absolute;
      top: -10.5rem;
      left: 50%;
      transform: translate(-50%);

      width: 10rem;
      height: 10rem;
      overflow: hidden;
      border-radius: 50%;
      box-shadow: 0rem 0rem 0.8rem var(--cl-primary);
    }

    &--img {
      width: 100%;
      height: 100%;
      display: block;
      object-fit: cover;
    }

    &--about {
      display: grid;
      grid-template-columns: min-content 1fr min-content;
    }

    &--qtstart {
      font-size: 6rem;
      align-self: start;
    }

    &--aboutctn {
      padding: 1rem;
    }

    &--qtend {
      font-size: 6rem;
      align-self: end;
    }

    &--password {
      align-self: end;
    }
  }
  &__controlls {
    display: grid;
    grid-template-columns: repeat(2, 1fr);
    grid-column-gap: 1rem;
    &--link {
      &:link,
      &:visited {
        display: grid;
        align-items: center;
        text-decoration: none;
        color: var(--cl-white);
        background-color: var(--cl-font);
        text-align: center;

        transition: background-color 0.3s;
      }

      &:hover,
      &:active {
        background-color: var(--cl-primary);
      }
    }
  }
}

*/
