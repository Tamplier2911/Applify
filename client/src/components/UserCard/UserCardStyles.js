import styled, { css } from "styled-components";
import { Link } from "react-router-dom";

const fieldSharedStyles = css`
  text-align: center;
  font-size: 1.8rem;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

export const UserCardContainer = styled.div`
  display: grid;
  grid-template-rows: 5fr 5fr 1fr;
  grid-row-gap: 1rem;
  font-family: var(--font-title);
  padding: 2rem;
  box-shadow: 0rem 0rem 0.8rem var(--cl-font);

  transition: box-shadow 0.3s;

  &:hover {
    box-shadow: 0rem 0rem 0.8rem var(--cl-primary);
  }
`;

export const UserCardHeader = styled.div`
  display: grid;
  grid-auto-rows: min-content;
`;

export const UserCardHeaderName = styled.div`
  ${fieldSharedStyles}
`;

export const UserCardHeaderEmail = styled.div`
  ${fieldSharedStyles}
`;

export const UserCardHeaderRole = styled.div`
  text-align: center;
  font-size: 1.8;
`;

export const UserCardBody = styled.div`
  position: relative;
  display: grid;
  grid-template-rows: min-content 1fr;
`;

export const UserCardBodyWrap = styled.div`
  position: absolute;
  top: -10.5rem;
  left: 50%;
  transform: translate(-50%);

  width: 10rem;
  height: 10rem;
  overflow: hidden;
  border-radius: 50%;
  box-shadow: 0rem 0rem 0.8rem var(--cl-primary);
`;

export const UserCardBodyImg = styled.img`
  width: 100%;
  height: 100%;
  display: block;
  object-fit: cover;
`;

export const UserCardBodyAbout = styled.div`
  display: grid;
  grid-template-columns: min-content 1fr min-content;
`;

export const UserCardBodyAboutLQT = styled.div`
  font-size: 6rem;
  align-self: start;
`;

export const UserCardBodyAboutCTN = styled.div`
  padding: 1rem;
  align-self: center;
  text-align: center;
`;

export const UserCardBodyAboutRQT = styled.div`
  font-size: 6rem;
  align-self: end;
`;

export const UserCardBodyPassword = styled.div`
  align-self: end;
`;

export const UserCardBodyPasswordOptions = styled.div``;

export const UserCardBodyPasswordChanged = styled.div``;

export const UserCardControlls = styled.div`
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  grid-column-gap: 1rem;
`;

export const UserCardControllsLink = styled(Link)`
  &:link,
  &:visited {
    display: grid;
    align-items: center;
    text-align: center;
    text-decoration: none;
    color: var(--cl-white);
    background-color: var(--cl-font);

    transition: background-color 0.3s;
  }

  &:hover,
  &:active {
    background-color: var(--cl-primary);
  }
`;
