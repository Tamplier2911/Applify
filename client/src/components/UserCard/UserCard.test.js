import React from "react";
import toJson from "enzyme-to-json";

// component
import UserCard from "./UserCard";

// utils
import {
  setShallowC,
  findByTestAttr,
  storeFactory,
  checkProps,
} from "../../tests/testUtils";

describe("<UserCard />", () => {
  const initialState = {
    lang: {
      languagePanelHidden: true,
      currentLanguage: "eng",
    },
  };
  const store = storeFactory(initialState);

  const userData = {
    about: "",
    email: "",
    likedBlogposts: [],
    name: "",
    passwordChangedAt: "",
    role: "",
    photo: "",
    _id: "",
  };

  let wrapper;
  beforeEach(() => {
    wrapper = setShallowC(<UserCard store={store} data={userData} />);
  });

  it("renders without an error", () => {
    const userCard = findByTestAttr(wrapper, "user-card");
    expect(userCard).toHaveLength(1);
  });

  it("renders children - header", () => {
    const userCardHeader = findByTestAttr(wrapper, "user-card-header");
    expect(userCardHeader).toHaveLength(1);
  });

  it("renders children - body", () => {
    const userCardBody = findByTestAttr(wrapper, "user-card-body");
    expect(userCardBody).toHaveLength(1);
  });

  it("renders children - controls", () => {
    const userCardControls = findByTestAttr(wrapper, "user-card-controls");
    expect(userCardControls).toHaveLength(1);
  });

  it("recieves expected props", () => {
    const expectedProps = {
      data: {
        about: "",
        email: "",
        likedBlogposts: [],
        name: "",
        passwordChangedAt: "",
        role: "",
        photo: "",
        _id: "",
      },
    };
    const err = checkProps(UserCard, expectedProps, "props");
    expect(err).toBeUndefined();
  });

  it("matches snapshot", () => {
    const userCard = findByTestAttr(wrapper, "user-card");
    expect(toJson(userCard)).toMatchSnapshot();
  });
});
