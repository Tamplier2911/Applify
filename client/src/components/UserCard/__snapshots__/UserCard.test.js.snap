// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserCard /> matches snapshot 1`] = `
<Styled(ScrollAnimation)
  animateIn="flipInY"
  animateOnce={true}
  data-test="user-card"
>
  <styled.div
    data-test="user-card-header"
  >
    <styled.div />
    <styled.div />
    <styled.div />
  </styled.div>
  <styled.div
    data-test="user-card-body"
  >
    <styled.div>
      <styled.img
        alt="happy user"
        src="http://localhost/api/uploads/images/users/default.jpg"
      />
    </styled.div>
    <styled.div>
      <styled.div>
        “
      </styled.div>
      <styled.div>
        About section is not yet modified.
      </styled.div>
      <styled.div>
        „
      </styled.div>
    </styled.div>
    <styled.div>
      <styled.div>
        Password was changed at:
      </styled.div>
      <styled.div>
        Password was not yet changed.
      </styled.div>
    </styled.div>
  </styled.div>
  <styled.div
    data-test="user-card-controls"
  >
    <Button
      click={[Function]}
      type="button"
      value="Delete"
    />
    <Styled(Link)
      to="/profile/users/"
    >
      Update
    </Styled(Link)>
  </styled.div>
</Styled(ScrollAnimation)>
`;
